packages/repository/src/repositories/relation.repository.ts

```diff
@@ -0,0 +1,173 @@
+ */
+export interface HasManyEntityCrudRepository<T extends Entity, ID>
+  extends Repository<T> {
+  /**
+   * Create a target model instance
+   * @param targetModelData The target model data
+   * @param options Options for the operation
+   * @returns A promise which resolves to the newly created target model instance
+   */
+  create(targetModelData: DataObject<T>, options?: Options): Promise<T>;
+
+  /**
+   * Build a target model instance
+   * @param targetModelData The target model data
+   * @returns A promise of the created model instance
+   */
+  build(targetModelData: DataObject<T>): Promise<T>;
+
+  /**
+   * Find a related entity by id
+   * @param id The foreign key
+   * @param options Options for the operation
+   * @returns A promise of an entity found for the id
+   */
+  findById(id: ID, filter?: Filter, options?: Options): Promise<T>;
+
+  /**
+   * Update a related entity by foreign key
+   * @param data Data attributes to be updated
+   * @param id Value for the foreign key
+   * @param options Options for the operation
+   * @returns Promise<true> if the entity is updated, otherwise
+   * Promise<false>
+   */
+  updateById(id: ID, data: DataObject<T>, options?: Options): Promise<boolean>;
+
+  /**
+   * Delete a related entity by id
+   * @param id Value for the entity foreign key
+   * @param options Options for the operation
+   * @returns Promise<true> if an entity is deleted for the id, otherwise
+   * Promise<false>
+   */
+  deleteById(id: ID, options?: Options): Promise<boolean>;
+
+  /**
+   * Check if the related entity exists for the given foreign key
+   * @param id Value for the entity foreign key
+   * @param options Options for the operation
+   * @returns Promise<true> if an entity exists for the id, otherwise
+   * Promise<false>
+   */
+  exists(id: ID, options?: Options): Promise<boolean>;
+}
+
+export class DefaultHasManyEntityCrudRepository<T extends Entity, ID>
+  implements HasManyEntityCrudRepository<T, ID> {
+  /**
+   * Constructor of DefaultCrudRepository
+   * @param targetRepo a repository of the target model
+   * @param foreignKey the constraints to scope target repo CRUD methods
+   */
+  constructor(
+    public targetRepo: DefaultCrudRepository<T, ID>,
+    public primaryKey: ID,
+  ) {}
+  execute(
+    command: string | AnyObject,
+    // tslint:disable-next-line:no-any
+    parameters: any[] | AnyObject,
+    options?: Options,
+  ): Promise<AnyObject> {
+    throw new Error('Method not implemented.');
+  }
+  /**
+   * Create a target model instance
+   * @param targetModelData The target model data
+   * @param options Options for the operation
+   * @returns A promise which resolves to the newly created target model instance
+   */
+  create(targetModelData: Partial<T>, options?: Options): Promise<T> {
+    options = Object.assign(options, this.buildFilter());
+    return this.targetRepo.create(targetModelData, options);
```

Comment:
This does not look correct to me.

Assuming "Product belongsTo Category" relation and a scope with categoryId = 1, I'd expect the following call to create a Product called "Pen":

```ts
this.targetRepo.create({name: 'Pen', categoryId: 1});
```

IIUC, your current code makes a different call, one that's not supported by juggler nor the juggler bridge AFAIK:

```ts
this.targetRepo.create({name: 'Pen'}, {categoryId: 1}});
```

I think this is exposing one source of the complexity of arbitrary-scoped repositories: we need the scope to provide us both a where-filter to use when reading data and a set of property values to include when writing to a database.

Additionally, we a scope may need to include a constraint on which scope-related properties are users allowed to change via updates.

For example, consider an application where a `Post` belongs to a `User`. Should we allow API clients to make a request that will reassign `Post` from one user to another? Imagine you posted something on Facebook or Twitter and then you could change the `userId` property to say the post was authored by me. 

On the other hand, there are valid use cases where it's ok to reassign "belongsTo" target. For example, if we have a Product called "Pen" belonging to Category "Stationery" with id 1, then it's probably ok to change pen's categoryId to 2 (a category called "Gifts").

@raymondfeng thoughts?

Reply:
@bajtos The scoped `create` needs to set `categoryId: 1` into the data. It's not `options`.

Reply:
Yes, I agree it was wrong before; going to re-work this :)